modules/nrfxlib/nrf_802154/sl/platform/nrf_802154_platform_sl_lptimer_zephyr.c

```diff
@@ -104,34 +117,134 @@ static bool hw_task_state_set(enum hw_task_state_type expected_state,
 
-static bool cc_event_check(int32_t cc_num)
+static bool hw_task_rtc_cc_event_check(int32_t cc_num)
 {
-	uint32_t event_address = z_nrf_rtc_timer_compare_evt_address_get(cc_num);
+	uint32_t event_address = hw_task_rtc_get_compare_evt_address(cc_num);
 
 	return *(volatile uint32_t *)event_address;
 }
 
+static void hw_task_rtc_timer_abort(void)
+{
+#if defined(CONFIG_SOC_NRF5340_CPUNET)
+	/* No interrupt disabling/clearing is needed, as HW task do not use interrupts*/
+	nrf_rtc_event_clear(HW_TASK_RTC, NRF_RTC_CHANNEL_EVENT_ADDR(m_hw_task.chan));
+	nrf_rtc_event_disable(HW_TASK_RTC, NRF_RTC_CHANNEL_INT_MASK(m_hw_task.chan));
+#else
+	z_nrf_rtc_timer_abort(m_hw_task.chan);
+#endif
+}
+
+#if defined(CONFIG_SOC_NRF5340_CPUNET)
+static uint32_t zephyr_rtc_ticks_to_hw_task_rtc_counter(uint64_t ticks)
+{
+	uint32_t zephyr_rtc_counter_now = 0;
+	uint32_t hw_task_rtc_counter_now = 0;
+	uint32_t hw_task_rtc_counter_now_2 = 0;
+
+	/**
+	 * The RTC0 counter is read twice - once before and once after reading RTC1.
+	 * If the two reads are different, the procedure is repeated.
+	 * This is to make sure that no RTC tick occured between reading RTC0 and
+	 * RTC1 and so the calculated difference between the timers is correct.
+	 */
+	do
+	{
+		hw_task_rtc_counter_now = nrf_rtc_counter_get(HW_TASK_RTC);
+		zephyr_rtc_counter_now = (uint32_t) (z_nrf_rtc_timer_read() % RTC_COUNTER_SPAN);
+		barrier_dmem_fence_full();
+		barrier_isync_fence_full();
+		hw_task_rtc_counter_now_2 = nrf_rtc_counter_get(HW_TASK_RTC);
+
+	} while(hw_task_rtc_counter_now != hw_task_rtc_counter_now_2);
+
+	uint32_t diff = (hw_task_rtc_counter_now + RTC_COUNTER_SPAN) - zephyr_rtc_counter_now;
+
+	return (uint32_t) ((ticks + diff) % RTC_COUNTER_SPAN);
+}
+
+static int hw_task_rtc_counter_compare_set(uint32_t cc_value)
+{
+	uint32_t current_counter_value = nrf_rtc_counter_get(HW_TASK_RTC);
+
+	/* As the HW tasks are setup only shortly before they are triggered,
+	 * it is certain that the required fire time is not further away
+	 * than one half of the RTC counter span.
+	 */
+	if(current_counter_value > cc_value)
+	{
+		if(current_counter_value - cc_value < RTC_COUNTER_HALF_SPAN)
+		{
+			return -EINVAL;
+		}
+	}
+
+	nrf_rtc_event_clear(HW_TASK_RTC, NRF_RTC_CHANNEL_EVENT_ADDR(m_hw_task.chan));
+	nrf_rtc_event_disable(HW_TASK_RTC, NRF_RTC_CHANNEL_INT_MASK(m_hw_task.chan));
+
+	nrf_rtc_cc_set(HW_TASK_RTC, m_hw_task.chan, cc_value & RTC_COUNTER_MAX);
+
+	nrf_rtc_event_enable(HW_TASK_RTC, NRF_RTC_CHANNEL_INT_MASK(m_hw_task.chan));
+
+	current_counter_value = nrf_rtc_counter_get(HW_TASK_RTC);
+	uint32_t time_to_event = (cc_value +  RTC_COUNTER_SPAN - current_counter_value) % RTC_COUNTER_SPAN;
+
+	if ( time_to_event < RTC_MIN_CYCLES_FROM_NOW || time_to_event > RTC_COUNTER_HALF_SPAN)
+	{
+		if(!hw_task_rtc_cc_event_check(m_hw_task.chan))
+		{
+			return -EINVAL;
+		}
+	}
+
+	return 0;
+}
+#endif
+
+static int hw_task_rtc_timer_set(uint64_t fire_lpticks)
+{
+#if defined(CONFIG_SOC_NRF5340_CPUNET)
```

Comment:
Missing equivalent of too distant time check using 64-bit arith.
reference https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/nrf_rtc_timer.c#L347

Missing equivalent of attempt to set in the past (far past) using 64-bit arith
reference https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/timer/nrf_rtc_timer.c#L361

Currently attempt to set `hw_task_rtc_timer_set(z_nrf_rtc_timer_read() - (RTC_COUNTER_SPAN - 10))` would result in a RTC scheduled 10 ticks in future, while fully 64-bit value is clearly in the past.

Naive fix would be to read `z_nrf_rtc_timer_read()` once more, perform checks and then if conditions are ok call `zephyr_rtc_ticks_to_hw_task_rtc_counter` and `hw_task_rtc_counter_compare_set`.

Better would be to change prototype of `zephyr_rtc_ticks_to_hw_task_rtc_counter`
```c
static bool zephyr_rtc_ticks_to_hw_task_rtc_counter(uint64_t ticks, uint32_t * cc_value)
```
that would validate internally `zephyr_rtc_counter_now` variable against `ticks` parameter
return true only if ticks are in future but not further than RTC_COUNTER_HALF_SPAN

Reply:
Agreed with @ahasztag. The HW tasks are always scheduled shortly before radio operation. Having half RTC span of margin is more than enough.